source/common/io/io_uring_worker_impl.cc

```diff
@@ -123,5 +275,340 @@ void IoUringWorkerImpl::submit() {
+      write_or_shutdown_cancel_req_ == nullptr) {
+    closeInternal();
+  }
+}
+
+void IoUringServerSocket::moveReadDataToBuffer(Request* req, size_t data_length) {
+  ReadRequest* read_req = static_cast<ReadRequest*>(req);
+  Buffer::BufferFragment* fragment = new Buffer::BufferFragmentImpl(
+      read_req->buf_.release(), data_length,
+      [](const void* data, size_t, const Buffer::BufferFragmentImpl* this_fragment) {
+        delete[] reinterpret_cast<const uint8_t*>(data);
+        delete this_fragment;
+      });
+  read_buf_.addBufferFragment(*fragment);
+}
+
+void IoUringServerSocket::onReadCompleted(int32_t result) {
+  ENVOY_LOG(trace, "read from socket, fd = {}, result = {}", fd_, result);
+  ReadParam param{read_buf_, result};
+  read_param_ = param;
+  IoUringSocketEntry::onReadCompleted();
+  read_param_ = absl::nullopt;
+  ENVOY_LOG(trace, "after read from socket, fd = {}, remain = {}", fd_, read_buf_.length());
+}
+
+// TODO(zhxie): concern submit multiple read requests or submit read request in advance to improve
+// performance in the next iteration.
+void IoUringServerSocket::onRead(Request* req, int32_t result, bool injected) {
+  IoUringSocketEntry::onRead(req, result, injected);
+
+  ENVOY_LOG(trace,
+            "onRead with result {}, fd = {}, injected = {}, status_ = {}, enable_close_event = {}",
+            result, fd_, injected, status_, enable_close_event_);
+  if (!injected) {
+    read_req_ = nullptr;
+    // If the socket is going to close, discard all results.
+    if (status_ == Closed && write_or_shutdown_req_ == nullptr && read_cancel_req_ == nullptr &&
+        write_or_shutdown_cancel_req_ == nullptr) {
+      if (result > 0 && keep_fd_open_) {
+        moveReadDataToBuffer(req, result);
+      }
+      closeInternal();
+      return;
+    }
+  }
+
+  // Move read data from request to buffer or store the error.
+  if (result > 0) {
+    moveReadDataToBuffer(req, result);
+  } else {
+    if (result != -ECANCELED) {
+      read_error_ = result;
+    }
+  }
+
+  // Discard calling back since the socket is not ready or closed.
+  if (status_ == Initialized || status_ == Closed) {
+    return;
+  }
+
+  // If the socket is enabled and there are bytes to read, notify the handler.
+  if (status_ == ReadEnabled) {
+    if (read_buf_.length() > 0) {
+      onReadCompleted(static_cast<int32_t>(read_buf_.length()));
+    } else if (read_error_.has_value() && read_error_ < 0) {
+      onReadCompleted(read_error_.value());
+      read_error_.reset();
+    }
+    // Handle remote closed at last.
+    // Depending on the event listened to, calling different event back to handle remote closed.
+    // * events & (Read | Closed): Callback Closed,
+    // * events & (Read)         : Callback Read,
+    // * events & (Closed)       : Callback Closed,
+    // * ...else                 : Callback Write.
+    if (read_error_.has_value() && read_error_ == 0 && !enable_close_event_) {
```

Comment:
I'm a bit confused by this logic. If `enable_close_event_` is true when the close event is reported as the even `Closed`. If it is false then it is reported as event `Read`. And in some conditions it can be reported as a `Write` event. 

What is the use case for all these different scenarios? Can we simplify and always report remote close event as a `Close` event?

Reply:
That is our ConnectionImpl's behavior.

We are only listening on read and write event in normal
https://github.com/envoyproxy/envoy/blob/7f84968b32bf33f93fb48e5fcfc3e14fbea95301/source/common/network/connection_impl.cc#L98-L102

We will listen the close event when the read is disabled
https://github.com/envoyproxy/envoy/blob/7f84968b32bf33f93fb48e5fcfc3e14fbea95301/source/common/network/connection_impl.cc#L423-L429

And we only listen write when half closed.